cpp/mrc/src/internal/service.cpp

```diff
@@ -17,131 +17,295 @@
 
 #include "internal/service.hpp"
 
+#include "mrc/core/utils.hpp"
+#include "mrc/exceptions/runtime_error.hpp"
+#include "mrc/utils/string_utils.hpp"
+
 #include <glog/logging.h>
 
+#include <future>
+#include <mutex>
 #include <ostream>
 #include <utility>
 
 namespace mrc {
 
+Service::Service(std::string service_name) : m_service_name(std::move(service_name)) {}
+
 Service::~Service()
 {
+    if (!m_call_in_destructor_called)
+    {
+        LOG(ERROR) << "Must call Service::call_in_destructor to ensure service is cleaned up before being "
+                      "destroyed";
+    }
+
     auto state = this->state();
     CHECK(state == ServiceState::Initialized || state == ServiceState::Completed);
 }
 
+const std::string& Service::service_name() const
+{
+    return m_service_name;
+}
+
+bool Service::is_service_startable() const
+{
+    std::lock_guard<decltype(m_mutex)> lock(m_mutex);
+    return (m_state == ServiceState::Initialized);
+}
+
+bool Service::is_running() const
+{
+    std::lock_guard<decltype(m_mutex)> lock(m_mutex);
+    return (m_state > ServiceState::Initialized && m_state < ServiceState::Completed);
+}
+
+const ServiceState& Service::state() const
+{
+    std::lock_guard<decltype(m_mutex)> lock(m_mutex);
+    return m_state;
+}
+
 void Service::service_start()
 {
-    if (forward_state(ServiceState::Running))
+    std::unique_lock<decltype(m_mutex)> lock(m_mutex);
+
+    if (!this->is_service_startable())
     {
-        do_service_start();
+        throw exceptions::MrcRuntimeError(MRC_CONCAT_STR(this->debug_prefix() << " Service has already been started"));
+    }
+
+    if (forward_state(ServiceState::Starting))
+    {
+        // Unlock the mutex before calling start to avoid a deadlock
+        lock.unlock();
+
+        try
+        {
+            this->do_service_start();
+
+            // Use ensure_state here in case the service itself called stop or kill
+            this->ensure_state(ServiceState::Running);
+        } catch (...)
+        {
+            // On error, set this to completed and rethrow the error to allow for cleanup
+            this->forward_state(ServiceState::Completed);
+
+            throw;
+        }
     }
 }
 
 void Service::service_await_live()
 {
-    do_service_await_live();
+    {
+        std::unique_lock<decltype(m_mutex)> lock(m_mutex);
+
+        if (this->is_service_startable())
```

Comment:
Remark: The inverse logic here ("because the service is not startable, the service must have been started already") does not follow logically unless the reader knows there is no state prior to `Initialized`, and that in no future case we add a state previous to `Initialized`. It's a bit confusing, but I'm not sure there's anything to be done about it.